File: localization/scripts/synchronizer.py
#!/usr/bin/env python3
import rospy
import message_filters
from sensor_msgs.msg import Imu, MagneticField
from geometry_msgs.msg import Vector3Stamped
from localization.msg import buffer

pub = rospy.Publisher("/imu_readings", buffer, queue_size=10)
rospy.init_node("sync_test")
r = rospy.Rate(10) #10hz
buffer_msg = buffer()

def sync_callback(imu_msg, mag_msg):
    # Create a buffer message
    buffer_msg.header.stamp = rospy.Time.now()
    # Assuming measurements is a list of 9 floats
    buffer_msg.measurements = imu_msg.angular_velocity.x, imu_msg.angular_velocity.y, imu_msg.angular_velocity.z, imu_msg.linear_acceleration.x, imu_msg.linear_acceleration.y, imu_msg.linear_acceleration.z, mag_msg.magnetic_field.x*1e-6, mag_msg.magnetic_field.y*1e-6, mag_msg.magnetic_field.z*1e-6
    # Publish the synchronized data
    print(buffer_msg)
    pub.publish(buffer_msg)

while not rospy.is_shutdown():
    # Create a publisher for synchronized data
    

    # Create subscribers for IMU and magnetometer data
    imu_sub = message_filters.Subscriber("/imu/data", Imu)
    mag_sub = message_filters.Subscriber("/imu/magnetometer", MagneticField)

    # Synchronize the messages from subscribers
    sync = message_filters.ApproximateTimeSynchronizer(
        [imu_sub, mag_sub], queue_size=5, slop=0.03)

    sync.registerCallback(sync_callback)
    rospy.loginfo(buffer_msg)

    r.sleep()
    # rospy.spin()

File: localization/scripts/synchronizer.cpp
#include <ros/ros.h>
#include <sensor_msgs/Imu.h>
#include <sensor_msgs/JointState.h>
#include <iostream>
#include <message_filters/subscriber.h>
#include <message_filters/synchronizer.h>
#include <message_filters/sync_policies/approximate_time.h>
#include <message_filters/time_synchronizer.h>
#include <ros/console.h>

using namespace message_filters;

class Node {
public:
    Node() : sub_imu(node, "imu", 10), sub_joint_states(node, "joint_states", 10), sync(sub_imu, sub_joint_states, 10) {
        sync.registerCallback(&Node::syncCallback, this);
    }

private:
    ros::NodeHandle node;
    message_filters::Subscriber<sensor_msgs::Imu> sub_imu;
    message_filters::Subscriber<sensor_msgs::JointState> sub_joint_states;
    TimeSynchronizer<sensor_msgs::Imu, sensor_msgs::JointState> sync;

    void imuCallback(const sensor_msgs::ImuConstPtr& msg) {
        // Process IMU data here
        ROS_INFO_STREAM("hi, Received IMU message at: " << msg->header.stamp);
    }

    void jointStatesCallback(const sensor_msgs::JointStateConstPtr& msg) {
        // Process joint states data here
        ROS_INFO_STREAM("hi, Received joint states message at: " << msg->header.stamp);
    }

    void syncCallback(const sensor_msgs::ImuConstPtr& imu_msg, const sensor_msgs::JointStateConstPtr& joint_states_msg) {
        // Check if time difference is within tolerance
        double time_diff = std::abs(imu_msg->header.stamp.toSec() - joint_states_msg->header.stamp.toSec());
        ROS_INFO_STREAM("HI");
        if (time_diff > 0.025) {
            ROS_INFO_STREAM("Time difference between IMU and joint states exceeds 0.03 seconds: " << time_diff);
            ROS_INFO_STREAM("Received IMU message sync at: " << imu_msg->header.stamp.nsec);
            ROS_INFO_STREAM("Received encoder message sync at: " << joint_states_msg->header.stamp.nsec);
            return; // Skip processing if time difference is too large
        }
        ROS_INFO_STREAM("Synchronized message at: " << time_diff);
        ROS_INFO_STREAM("Received IMU message synv at: " << imu_msg->header.stamp.nsec);
        ROS_INFO_STREAM("Received encoder message synv at: " << joint_states_msg->header.stamp.nsec);
    }
};

int main(int argc, char **argv) {
    ros::init(argc, argv, "synchronizer");
    Node node;
    
    ros::spin();
    return 0;
}
